include/xparrow/xbuffer.hpp

```diff
@@ -0,0 +1,239 @@
+        {
+            return size() == size_type(0);
+        }
+
+        template <class T>
+        auto xbuffer_data<T>::size() const noexcept -> size_type
+        {
+            return m_size;
+        }
+
+        template <class T>
+        template <class U>
+        U* xbuffer_data<T>::data() noexcept
+        {
+            return reinterpret_cast<U*>(p_data);
+        }
+
+        template <class T>
+        template <class U>
+        const U* xbuffer_data<T>::data() const noexcept
+        {
+            return reinterpret_cast<const U*>(p_data);
+        }
+
+        template <class T>
+        void xbuffer_data<T>::swap(xbuffer_data<T>& rhs) noexcept
+        {
+            std::swap(p_data, rhs.p_data);
+            std::swap(m_size, rhs.m_size);
+        }
+
+        template <class T>
+        bool xbuffer_data<T>::equal(const xbuffer_data<T>& rhs) const
+        {
+            return m_size == rhs.m_size && std::equal(p_data, p_data + m_size, rhs.p_data);
+        }
+    }
+
+    /*************************
+     * xbuffer implementation *
+     *************************/
+
+    template <class T>
+    xbuffer<T>::xbuffer(size_type size)
+        : base_type{allocate(size), size}
+    {
+    }
+
+    template <class T>
+    xbuffer<T>::xbuffer(pointer data, size_type size)
+        : base_type{data, size}
+    {
+    }
+        
+    template <class T>
+    xbuffer<T>::~xbuffer()
+    {
+        deallocate(this->p_data);
+        this->p_data = nullptr;
+        this->m_size = 0u;
+    }
+        
+    template <class T>
+    xbuffer<T>::xbuffer(const xbuffer<T>& rhs)
+        : base_type{allocate(rhs.m_size), rhs.size()}
+    {
+        std::copy(rhs.data(), rhs.data() + rhs.size(), data());
+    }
+
+    template <class T>
+    xbuffer<T>& xbuffer<T>::operator=(const xbuffer<T>& rhs)
+    {
+        if (this != &rhs)
+        {
+            xbuffer<T> tmp(rhs);
+            swap(tmp);
+        }
+        return *this;
+    }
+
+    template <class T>
+    xbuffer<T>::xbuffer(xbuffer&& rhs)
+        : base_type{rhs.data(), rhs.size()}
+    {
+        rhs.p_data = nullptr;
+        rhs.m_size = 0u;
+    }
+
+    template <class T>
+    xbuffer<T>& xbuffer<T>::operator=(xbuffer<T>&& rhs)
+    {
+        swap(rhs);
+        return *this;
+    }
+
+    template <class T>
+    void xbuffer<T>::swap(xbuffer<T>& rhs) noexcept
+    {
+        base_type::swap(rhs);
+    }
+
+    template <class T>
+    bool xbuffer<T>::equal(const xbuffer<T>& rhs) const
+    {
+        return base_type::equal(rhs);
+    }
+
+    template <class T>
+    auto xbuffer<T>::allocate(size_type size) const -> pointer
+    {
+        return new T[size];
+    }
+
+    template <class T>
+    void xbuffer<T>::deallocate(pointer mem) const
+    {
+        delete[] mem;
+    }
+
+    template <class T>
+    bool operator==(const xbuffer<T>& lhs, const xbuffer<T>& rhs)
+    {
+        return lhs.equal(rhs);
+    }
```

Comment:
This is required by doctest.
```suggestion
    template <class T>
    bool operator==(const xbuffer<T>& lhs, const xbuffer<T>& rhs)
    {
        return lhs.equal(rhs);
    }
    
    template <class T>
    bool operator!=(const xbuffer<T>& lhs, const xbuffer<T>& rhs)
    {
        return !(lhs == rhs);
    }
```

Reply:
Excuse my interjection, this is not needed anymore with C++20.